panther_manager/src/manager_bt_node.cpp

```diff
@@ -57,48 +195,71 @@ void ManagerNode::e_stop_cb(const std_msgs::Bool::ConstPtr & e_stop)
+}
+
 void ManagerNode::lights_tree_timer_cb()
 {
   // update blackboard
   lights_config_.blackboard->set<bool>("e_stop_state", e_stop_state_.value());
-  lights_config_.blackboard->set<bool>("charging_state", io_state_.value()->charger_connected);
+  lights_config_.blackboard->set<unsigned>("battery_status", battery_status_.value());
+  lights_config_.blackboard->set<float>("battery_percent", battery_percent_);
+  lights_config_.blackboard->set<std::string>(
+    "battery_percent_round",
+    std::to_string(
+      round(battery_percent_ / update_charging_anim_step_) * update_charging_anim_step_));
 
-  // BT::StdCoutLogger logger_cout(lights_tree_);  // debuging
+  // BT::StdCoutLogger logger_cout(lights_tree_);  // debugging
   lights_tree_status_ = lights_tree_.tickOnce();
 }
 
 void ManagerNode::safety_tree_timer_cb()
 {
   // update blackboard
-  safety_config_.blackboard->set<bool>("fan_state", io_state_.value()->fan);
   safety_config_.blackboard->set<bool>("aux_state", io_state_.value()->aux_power);
+  safety_config_.blackboard->set<bool>("e_stop_state", e_stop_state_.value());
+  safety_config_.blackboard->set<bool>("fan_state", io_state_.value()->fan);
+  safety_config_.blackboard->set<double>("bat_temp", battery_temp_ma_.get_average());
+  safety_config_.blackboard->set<double>("cpu_temp", cpu_temp_ma_.get_average());
+  // to simplify conditions pass only higher temp of motor drivers
+  safety_config_.blackboard->set<double>(
+    "driver_temp",
+    std::max({front_driver_temp_ma_.get_average(), rear_driver_temp_ma_.get_average()}));
 
-  // BT::StdCoutLogger logger_cout(safety_tree_);  // debuging
+  // BT::StdCoutLogger logger_cout(safety_tree_);  // debugging
   safety_tree_status_ = safety_tree_.tickOnce();
+
+  std::string signal_shutdown;
+  if (safety_config_.blackboard->get<std::string>("signal_shutdown", signal_shutdown)) {
+    shutdown_robot(signal_shutdown);
+  }
 }
 
-void ManagerNode::setup_behavior_tree(
-  BT::Tree & tree, BT::NodeConfig & config, const std::string tree_name)
+void ManagerNode::shutdown_robot(const std::string & message)
 {
-  config.blackboard = BT::Blackboard::create();
-  config.blackboard->set("nh", nh_);
-  tree = factory_.createTree(tree_name, config.blackboard);
+  ROS_WARN("[%s] %s. Soft shutdown initialized.", node_name_.c_str(), message.c_str());
+  lights_tree_timer_.stop();
+  lights_tree_.haltTree();
+  safety_tree_timer_.stop();
+  safety_tree_.haltTree();
+
+  // tick shutdown tree
+  // BT::StdCoutLogger logger_cout(shutdown_tree_);  // debugging
+  shutdown_tree_status_ = BT::NodeStatus::RUNNING;
+  auto start_time = ros::Time::now();
+  while (ros::ok() && shutdown_tree_status_ == BT::NodeStatus::RUNNING) {
+    auto shutdown_timeout = (ros::Time::now() - start_time) > ros::Duration(shutdown_timeout_);
+    shutdown_config_.blackboard->set<bool>("shutdown_timeout", shutdown_timeout);
+    shutdown_tree_status_ = shutdown_tree_.tickOnce();
+  }
```

Comment:
You used in this case is `ros::Rate rate(10.0)`, while in constructor you use `ros::Duration(0.1).sleep();` I think it would be better if in both cases the same approach was used